include/openrave/kinbody.h

```diff
@@ -3586,6 +3590,8 @@ class OPENRAVE_API KinBody : public InterfaceBase
     }
 
 protected:
+    using MapGrabbedByEnvironmentIndex = std::unordered_map<size_t, GrabbedPtr>;
```

Comment:
Why size_t? Shouldn't it be int?

Reply:
what would be the impact if this was an ordered std::vector?

Reply:
Re: performance impact, here's a test sweep:
https://gist.github.com/rschlaikjer/a20a0c846fa832f1a24805c531a2612b

Low hit rate:
![2024-08-27-103532_3836x2114_scrot](https://github.com/user-attachments/assets/1edbc3d5-4305-44ed-8d1b-936e7716c93c)

High hit rate:
![2024-08-27-103459_3836x2114_scrot](https://github.com/user-attachments/assets/06039512-6fec-4727-b0a1-46176b243fdb)

The sorted vector honestly does way better than I expected for search, but as expected, as N gets large you cannot escape the computational complexity of each implementation.
If we focus in on the low-N region, things are pretty noisy, but it doesn't _seem_ like there's a clear advantage to using a  vector:

![2024-08-27-104358_3836x2114_scrot](https://github.com/user-attachments/assets/8b3d6a40-0be2-4d56-85aa-0ceb9642afef)
![2024-08-27-104327_3836x2114_scrot](https://github.com/user-attachments/assets/dc81aa7b-5afb-4abc-9617-bafb9b622384)

Reply:
Thanks for the experiments. Two comments:

1. Let's stick with experimenting with realistic numbers. For example, N will be at max 100
2. Your test case has an extra indirection with std vector (datum->data_id < search), which does not exist in the map case. In order to make this apples-to-apples test, please manage the vector elements as a pair, with first being the key. Also, probably best to use int as the key, since that is what the openrave code will be doing

Thanks

Reply:
Interestingly, pair seems to be a lot worse - possible there's a mistake in my benchmark? For non SSO strings I wouldn't expect storing them as a pair to help with locality since the data is heaped somewhere else anyway, so you have the same dereference you'd have just looking at the Data member. For ints... I'm not sure why it's so much worse than the others. 

https://gist.github.com/rschlaikjer/ddcf2d71458f83de9e569c92bb47a8d6

![2024-08-28-091931_3836x2114_scrot](https://github.com/user-attachments/assets/fcf2c4be-4d5e-49fd-b31e-169faec2a90c)
![2024-08-28-091858_3836x2114_scrot](https://github.com/user-attachments/assets/a44fc6d7-52a4-496f-b156-271fd727e5ce)

plotting - 
```R
map_data <- read.csv("map_results.csv")
vec_data <- read.csv("vec_results.csv")
vec_pair_data <- read.csv("vec_pair_results.csv")
par(mfrow=c(2,1))

plot(vec_data$haystack_size, vec_data$insert, col='blue', type='p', ylim=c(0, 1800))
lines(map_data$haystack_size, map_data$insert, col='green', type='p')
lines(vec_pair_data$haystack_size, vec_pair_data$insert, col='orange', type='p')
legend(90, 1500, legend=c("vector", "map", "vec_pair"), fill=c("blue", "green", "orange"))
title("Amortized insert time (ns) vs element count (90% hit rate, string32 key)")

plot(vec_data$haystack_size, vec_data$query, col='blue', type='p', ylim=c(0, 1200))
lines(map_data$haystack_size, map_data$query, col='green', type='p')
lines(vec_pair_data$haystack_size, vec_pair_data$insert, col='orange', type='p')
legend(90, 1200, legend=c("vector", "map", "vec_pair"), fill=c("blue", "green", "orange"))
title("Amortized query time (ns) vs element count (90% hit rate, string32 key)")
```

Reply:
Okay, thanks for the experiments. For now, looks like unordered_map wins. Please update MINOR version and CHANGELOG to get ready for merge.

Reply:
It looks like the pessimal insert is caused by all the shuffling the vector has to do - moving a pair is surprisingly expensive compared to just moving the pointer.